PWGHF/HFC/TableProducer/HFCorrelatorD0Hadrons.cxx

```diff
@@ -0,0 +1,491 @@
+      double efficiencyWeight = 1.;
+      if (flagApplyEfficiency) {
+        efficiencyWeight = 1. / efficiencyDmeson->at(o2::analysis::findBin(bins, candidate1.pt()));
+      }
+      // ========================== Fill mass histo  ================================
+      if (candidate1.isSelD0() >= selectionFlagD0) {
+        registry.fill(HIST("hMass"), InvMassD0(candidate1), candidate1.pt(), efficiencyWeight);
+        registry.fill(HIST("hMass1D"), InvMassD0(candidate1), efficiencyWeight);
+        registry.fill(HIST("hMassD01D"), InvMassD0(candidate1), efficiencyWeight);
+      }
+      if (candidate1.isSelD0bar() >= selectionFlagD0bar) {
+        registry.fill(HIST("hMass"), InvMassD0bar(candidate1), candidate1.pt(), efficiencyWeight);
+        registry.fill(HIST("hMass1D"), InvMassD0bar(candidate1), efficiencyWeight);
+        registry.fill(HIST("hMassD0bar1D"), InvMassD0bar(candidate1), efficiencyWeight);
+      }
+      // ========================== Fill general histos ================================
+      registry.fill(HIST("hPtCand"), candidate1.pt());
+      registry.fill(HIST("hPtProng0"), candidate1.ptProng0());
+      registry.fill(HIST("hPtProng1"), candidate1.ptProng1());
+      registry.fill(HIST("hEta"), candidate1.eta());
+      registry.fill(HIST("hPhi"), candidate1.phi());
+      registry.fill(HIST("hY"), YD0(candidate1));
+      registry.fill(HIST("hSelectionStatus"), candidate1.isSelD0bar() + (candidate1.isSelD0() * 2));
+
+      // ================================================================================= D-h correlation dedicated section =====================================================
+
+      // ========================== track loop starts here ================================
+      for (const auto& track : tracks) {
+        registry.fill(HIST("hTrackCounterData"), 1); // fill total no. of tracks
+        // Remove D0 daughters by checking track indices
+        if ((candidate1.index0Id() == track.mRowIndex) || (candidate1.index1Id() == track.mRowIndex)) {
+          continue;
+        }
+        if (std::abs(track.dcaXY()) >= 1. || std::abs(track.dcaZ()) >= 1.)
+          continue; // Remove secondary tracks
+
+        registry.fill(HIST("hTrackCounterData"), 2); // fill no. of tracks before soft pion removal
+
+        // ===== soft pion removal ===================================================
+        double InvMassDstar1 = 0., InvMassDstar2 = 0.;
+        bool isSoftpiD0 = false, isSoftpiD0bar = false;
+        double psum2 = RecoDecay::p2((candidate1.px() + track.px()), (candidate1.py() + track.py()), (candidate1.pz() + track.pz()));
+        double Epion = std::sqrt((track.px() * track.px()) + (track.py() * track.py()) + (track.pz() * track.pz()) + (massPi * massPi));
+        InvMassDstar1 = RecoDecay::m(std::sqrt(psum2), (EPiProng0 + EKProng1 + Epion));
+        InvMassDstar2 = RecoDecay::m(std::sqrt(psum2), (EKProng0 + EPiProng1 + Epion));
```

Comment:
I'm still wondering why you do so many calculations by hand here and on lines 170-173, when it seems that what you eventually want to do is just to call `RecoDecay::m` twice with different mass hypotheses for the D0-candidate prongs.

Reply:
If you want to avoid repeating certain calculations, you can calculate the two D0-candidate energy hypotheses beforehand. But I don't see the benefit of doing the other calculations by hand.

Reply:
Ciao Vit, here the goal is calculating a 3-particle invariant mass value, considering he two daughters of the D0 (under the two possible mass assumptions, which lead to different daughter energies) and, as third particle, the track that is being looped over.
Are you aware of quicker ways to do this invariant mass calculation with the ingredients already available?
This would be possible if dedicated methods to get the D0 energy from the sum of the daughter energies, under the two mass hypotheses, existed, but this doesn't seem to be present (ED0 just gives the D0 energy starting from the D0mass and D0 momentum).

For the rest, the pion energy calculation can be semplified a bit using track.p() instead of using the components, or even directly using the track.energy() method.
And you could use directly the RecoDecay::p method instead of RecoDecay::p2, to avoid the sqrt(psum2) operations

Reply:
Hi @fcolamar , thanks for your comment. Indeed, given the available tools, I would suggest the following steps to avoid unnecessary `sqrt`:
```c++
// outside the track loop
auto ePiK = RecoDecay::e(candidate1.pVectorProng0(), massPi) + RecoDecay::e(candidate1.pVectorProng1(), massK);
auto eKPi = RecoDecay::e(candidate1.pVectorProng0(), massK) + RecoDecay::e(candidate1.pVectorProng1(), massPi);
// inside the track loop
auto pSum2 = RecoDecay::p2(candidate1.px() + track.px(), candidate1.py() + track.py(), candidate1.pz() + track.pz());
auto ePion = track.energy(massPi);
invMassDstar1 = std::sqrt(pSum2, ePiK + ePion);
invMassDstar2 = std::sqrt(pSum2, eKPi + ePion);
```
What do you think?

Reply:
Hi @vkucera, yes it's also what I had in mind, it should be equivalent to what previously done, but more compact and efficient.

Reply:
Thanks @fcolamar and @vkucera for the suggestions, I have implemented them.